### ISSUE TITLE
@Any and @ManyToAny Associations Fail in Spring Boot 3

### ISSUE THREAD

**Issue by dirkniblick** [spring-projects/spring-data-jpa#2935]
I have polymorphic relationships setup between some of the entities in a Spring Boot project for my employer. I recently updated the project to Spring Boot 3 so I could start using Hibernate 6's improved custom collections. That's working perfectly, but now I'm getting what have to be regression failures when trying to [perist ANY polymorphic entities](https://docs.jboss.org/hibernate/orm/6.2/userguide/html_single/Hibernate_User_Guide.html#associations-any):

```
Cannot invoke "org.hibernate.metamodel.mapping.AttributeMapping.getAttributeMetadata()" because the return value of "org.hibernate.sql.results.graph.Fetchable.asAttributeMapping()" is null
java.lang.NullPointerException: Cannot invoke "org.hibernate.metamodel.mapping.AttributeMapping.getAttributeMetadata()" because the return value of "org.hibernate.sql.results.graph.Fetchable.asAttributeMapping()" is null
	at org.hibernate.loader.ast.internal.LoaderSelectBuilder.lambda$createFetchableBiConsumer$3(LoaderSelectBuilder.java:858)
	at org.hibernate.loader.ast.internal.LoaderSelectBuilder.visitFetches(LoaderSelectBuilder.java:764)
	at org.hibernate.loader.ast.internal.LoaderSqlAstCreationState.visitFetches(LoaderSqlAstCreationState.java:133)
	at org.hibernate.metamodel.mapping.internal.DiscriminatedAssociationMapping$AnyValuedResultGraphNode.afterInitialize(DiscriminatedAssociationMapping.java:382)
	at org.hibernate.metamodel.mapping.internal.DiscriminatedAssociationMapping$AnyValuedFetch.<init>(DiscriminatedAssociationMapping.java:501)
	at org.hibernate.metamodel.mapping.internal.DiscriminatedAssociationMapping.generateFetch(DiscriminatedAssociationMapping.java:339)
	at org.hibernate.metamodel.mapping.internal.DiscriminatedAssociationAttributeMapping.generateFetch(DiscriminatedAssociationAttributeMapping.java:138)
	at org.hibernate.sql.results.graph.FetchParent.generateFetchableFetch(FetchParent.java:108)
	at org.hibernate.loader.ast.internal.LoaderSelectBuilder.lambda$createFetchableBiConsumer$3(LoaderSelectBuilder.java:933)
	at org.hibernate.loader.ast.internal.LoaderSelectBuilder.visitFetches(LoaderSelectBuilder.java:764)
	at org.hibernate.loader.ast.internal.LoaderSqlAstCreationState.visitFetches(LoaderSqlAstCreationState.java:133)
	at org.hibernate.sql.results.graph.AbstractFetchParent.afterInitialize(AbstractFetchParent.java:32)
	at org.hibernate.sql.results.graph.entity.AbstractEntityResultGraphNode.afterInitialize(AbstractEntityResultGraphNode.java:65)
	at org.hibernate.persister.entity.AbstractEntityPersister.createDomainResult(AbstractEntityPersister.java:1204)
	at org.hibernate.loader.ast.internal.LoaderSelectBuilder.generateSelect(LoaderSelectBuilder.java:467)
	at org.hibernate.loader.ast.internal.LoaderSelectBuilder.createSelect(LoaderSelectBuilder.java:256)
	at org.hibernate.loader.ast.internal.SingleIdEntityLoaderStandardImpl.createLoadPlan(SingleIdEntityLoaderStandardImpl.java:180)
	at org.hibernate.loader.ast.internal.SingleIdEntityLoaderStandardImpl.resolveLoadPlan(SingleIdEntityLoaderStandardImpl.java:120)
	at org.hibernate.loader.ast.internal.SingleIdEntityLoaderStandardImpl.load(SingleIdEntityLoaderStandardImpl.java:66)
	at org.hibernate.persister.entity.AbstractEntityPersister.doLoad(AbstractEntityPersister.java:3381)
	at org.hibernate.persister.entity.AbstractEntityPersister.load(AbstractEntityPersister.java:3371)
	at org.hibernate.event.internal.DefaultLoadEventListener.loadFromDatasource(DefaultLoadEventListener.java:602)
	at org.hibernate.event.internal.DefaultLoadEventListener.loadFromCacheOrDatasource(DefaultLoadEventListener.java:588)
	at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:557)
	at org.hibernate.event.internal.DefaultLoadEventListener.doLoad(DefaultLoadEventListener.java:550)
	at org.hibernate.event.internal.DefaultLoadEventListener.load(DefaultLoadEventListener.java:202)
	at org.hibernate.event.internal.DefaultLoadEventListener.loadWithRegularProxy(DefaultLoadEventListener.java:282)
	at org.hibernate.event.internal.DefaultLoadEventListener.proxyOrLoad(DefaultLoadEventListener.java:237)
	at org.hibernate.event.internal.DefaultLoadEventListener.doOnLoad(DefaultLoadEventListener.java:106)
	at org.hibernate.event.internal.DefaultLoadEventListener.onLoad(DefaultLoadEventListener.java:78)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:138)
	at org.hibernate.internal.SessionImpl.fireLoadNoChecks(SessionImpl.java:1231)
	at org.hibernate.internal.SessionImpl.fireLoad(SessionImpl.java:1219)
	at org.hibernate.loader.internal.IdentifierLoadAccessImpl.doLoad(IdentifierLoadAccessImpl.java:194)
	at org.hibernate.loader.internal.IdentifierLoadAccessImpl.lambda$load$1(IdentifierLoadAccessImpl.java:160)
	at org.hibernate.loader.internal.IdentifierLoadAccessImpl.perform(IdentifierLoadAccessImpl.java:107)
	at org.hibernate.loader.internal.IdentifierLoadAccessImpl.load(IdentifierLoadAccessImpl.java:160)
	at org.hibernate.internal.SessionImpl.get(SessionImpl.java:1024)
	at org.hibernate.event.internal.DefaultMergeEventListener.lambda$entityIsDetached$0(DefaultMergeEventListener.java:321)
	at org.hibernate.engine.spi.LoadQueryInfluencers.fromInternalFetchProfile(LoadQueryInfluencers.java:79)
	at org.hibernate.event.internal.DefaultMergeEventListener.entityIsDetached(DefaultMergeEventListener.java:319)
	at org.hibernate.event.internal.DefaultMergeEventListener.merge(DefaultMergeEventListener.java:150)
	at org.hibernate.event.internal.DefaultMergeEventListener.doMerge(DefaultMergeEventListener.java:143)
	at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:127)
	at org.hibernate.event.internal.DefaultMergeEventListener.onMerge(DefaultMergeEventListener.java:81)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.fireMerge(SessionImpl.java:848)
	at org.hibernate.internal.SessionImpl.merge(SessionImpl.java:834)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:311)
	at jdk.proxy3/jdk.proxy3.$Proxy103.merge(Unknown Source)
	at org.springframework.data.jpa.repository.support.SimpleJpaRepository.save(SimpleJpaRepository.java:613)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker$RepositoryFragmentMethodInvoker.lambda$new$0(RepositoryMethodInvoker.java:288)
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker.doInvoke(RepositoryMethodInvoker.java:136)
	at org.springframework.data.repository.core.support.RepositoryMethodInvoker.invoke(RepositoryMethodInvoker.java:120)
	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:516)
	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:285)
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:628)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.doInvoke(QueryExecutorMethodInterceptor.java:168)
	at org.springframework.data.repository.core.support.QueryExecutorMethodInterceptor.invoke(QueryExecutorMethodInterceptor.java:143)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:77)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:391)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:137)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:163)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:223)
	at jdk.proxy3/jdk.proxy3.$Proxy110.save(Unknown Source)
	at com.example.hibernatepolymorph.HibernatePolymorphApplicationTests.createPropertyHolder(HibernatePolymorphApplicationTests.java:55)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
...
```

In an effort to ensure it was not _my_ code causing this problem, I created a blank Spring Boot 3.0.6 project (with Hibernate 6.1.7.Final), added [the exact examples in the Hibernate documentation](https://docs.jboss.org/hibernate/orm/6.2/userguide/html_single/Hibernate_User_Guide.html#associations-any-property-example), and recreated the error on persisting either an @Any or @ManyToAny entity.

I forced the Spring Boot 3 project to use both Hibernate 6.0.2.Final and 6.2.2.Final, but they both gave me the same `Fetchable.asAttributeMapping() is null` problem. I also permuted the versions with Spring Boot 3.0.0, but that gave a different problem:

```
class org.hibernate.metamodel.mapping.internal.AnyDiscriminatorPart cannot be cast to class org.hibernate.metamodel.mapping.AttributeMapping (org.hibernate.metamodel.mapping.internal.AnyDiscriminatorPart and org.hibernate.metamodel.mapping.AttributeMapping are in unnamed module of loader 'app')
java.lang.ClassCastException: class org.hibernate.metamodel.mapping.internal.AnyDiscriminatorPart cannot be cast to class org.hibernate.metamodel.mapping.AttributeMapping (org.hibernate.metamodel.mapping.internal.AnyDiscriminatorPart and org.hibernate.metamodel.mapping.AttributeMapping are in unnamed module of loader 'app')
```

I then duplicated the same project, swapped it to Spring Boot 2.6.11 (with Hibernate 5.6.15.Final), updated the example classes to fit the previous method for setting up polymorphic relationships, and persisting polymorphic entities **worked immediately**.

I posted this problem [on the Hibernate forums](https://discourse.hibernate.org/t/any-and-manytoany-associations-fail-in-spring-boot-3/7576) and they asked me to reproduce it without using Spring Boot. So, I replaced Spring Boot with the same version of Hibernate used by 3.0.6 and the same entities persisted fine.

Below are links to the three projects, but I can't get an entity with an @Any or @ManyToAny annotation to work with Hibernate 6 on Spring Boot 3. Can anyone tell me why?

* https://github.com/dirkniblick/polymorphOnSpringBoot3
* https://github.com/dirkniblick/polymorphOnSpringBoot2
* https://github.com/dirkniblick/polymorphWithoutSpringboot

This is all you need to do to reproduce the failure using the same Java 17 SDK:

```
git clone https://github.com/dirkniblick/polymorphOnSpringBoot3.git
cd polymorphOnSpringBoot3/
gradle test --info
```

You will see the first one fails with the error below while the other two pass the same tests:

```
> Task :test FAILED

HibernatePolymorphApplicationTests > createPropertyRepository() FAILED
    java.lang.NullPointerException: Cannot invoke "org.hibernate.metamodel.mapping.AttributeMapping.getAttributeMetadataAccess()" because the return value of "org.hibernate.sql.results.graph.Fetchable.asAttributeMapping()" is null
```

**Comment by dirkniblick** [comment]
I've been doing loads of troubleshooting and debugging over the last few days. If the entities failing to be persisted are instead created **without** an id like this...
```java
        PropertyHolder namePropertyHolder = new PropertyHolder();
//        namePropertyHolder.setId(1L);
        namePropertyHolder.setProperty(nameProperty);

        propertyHolderRepository.saveAndFlush(namePropertyHolder);
        System.out.printf("Created: %s%n", namePropertyHolder);
```
...or this...
```java
        PropertyRepository propertyRepository = new PropertyRepository();
//        propertyRepository.setId(1L);

        propertyRepository.getProperties().add(ageProperty);
        propertyRepository.getProperties().add(nameProperty);

        propertyRepository = propertyRepositoryRepository.saveAndFlush(propertyRepository);
        System.out.printf("Created: %s%n", propertyRepository);
```
...it works, regardless of whether you add any related entities. But, **if the entity has an id, it will fail**. So, if you create one with an id...
```java
        PropertyHolder namePropertyHolder = new PropertyHolder();
        namePropertyHolder.setId(1L);
        namePropertyHolder.setProperty(nameProperty);

        propertyHolderRepository.saveAndFlush(namePropertyHolder);
        System.out.printf("Created: %s%n", namePropertyHolder);
```
...or, worse yet, try to update an existing entity...
```java
        PropertyRepository propertyRepository = new PropertyRepository();
//        propertyRepository.setId(1L);

        propertyRepository.getProperties().add(ageProperty);
        propertyRepository.getProperties().add(nameProperty);

        propertyRepository = propertyRepositoryRepository.saveAndFlush(propertyRepository); // This works fine.
        System.out.printf("Created: %s%n", propertyRepository);
...
        propertyRepository.getProperties().add(ageProperty2);

        propertyRepository = propertyRepositoryRepository.saveAndFlush(propertyRepository); // Throws NullPointerException
```
...it will throw the NullPointerException **every** time. This makes migration to Spring Boot 3 impossible for anyone already using Hibernate's @Any or @ManyToAny relationships.

**Comment by dirkniblick** [comment]
FYI, my communications with the Hibernate developers indicate [this is a bug with Hibernate 3](https://discourse.hibernate.org/t/any-and-manytoany-associations-fail-in-spring-boot-3/7576/20), not Spring Boot. They related this to one of [their Jira tickets](https://hibernate.atlassian.net/browse/HHH-16532) and it looks like [they fixed it in Hibernate 6.2.4.Final](https://in.relation.to/2023/06/01/hibernate-orm-624-final/#null-pointer-exceptions).

I updated [my test case](https://github.com/dirkniblick/polymorphOnSpringBoot3) with...
```
implementation 'org.hibernate.orm:hibernate-core:6.2.4.Final'
```
...and it works. I also updated it to use Spring Boot 3.1.1, which uses Hibernate 6.2.5.Final, and it works too.

**Comment by gregturn** [comment]
@dirkniblick Glad they were able to patch it. I'm now closing this issue.
